src/ln/channelmanager.rs

```diff
@@ -2077,152 +2121,92 @@ impl ChannelManager {
-										let update_len = byte_utils::slice_to_be16(&err_packet.failuremsg[offset+2..offset+4]) as usize;
-										if err_packet.failuremsg.len() >= offset + 4 + update_len {
-											if let Ok(chan_update) = msgs::ChannelUpdate::read(&mut Cursor::new(&err_packet.failuremsg[offset + 4..offset + 4 + update_len])) {
-												// if channel_update should NOT have caused the failure:
-												// MAY treat the channel_update as invalid.
-												let is_chan_update_invalid = match error_code {
-													c if c == UPDATE|7 => { // temporary_channel_failure
-														false
-													},
-													c if c == UPDATE|11 => { // amount_below_minimum
-														let reported_htlc_msat = byte_utils::slice_to_be64(&err_packet.failuremsg[2..2+8]);
-														onion_failure_log!("amount_below_minimum", UPDATE|11, "htlc_msat", reported_htlc_msat);
-														incoming_htlc_msat > chan_update.contents.htlc_minimum_msat
-													},
-													c if c == UPDATE|12 => { // fee_insufficient
-														let reported_htlc_msat = byte_utils::slice_to_be64(&err_packet.failuremsg[2..2+8]);
-														let new_fee =  amt_to_forward.checked_mul(chan_update.contents.fee_proportional_millionths as u64).and_then(|prop_fee| { (prop_fee / 1000000).checked_add(chan_update.contents.fee_base_msat as u64) });
-														onion_failure_log!("fee_insufficient", UPDATE|12, "htlc_msat", reported_htlc_msat);
-														new_fee.is_none() || incoming_htlc_msat >= new_fee.unwrap() && incoming_htlc_msat >= amt_to_forward + new_fee.unwrap()
-													}
-													c if c == UPDATE|13 => { // incorrect_cltv_expiry
-														let reported_cltv_expiry = byte_utils::slice_to_be32(&err_packet.failuremsg[2..2+4]);
-														onion_failure_log!("incorrect_cltv_expiry", UPDATE|13, "cltv_expiry", reported_cltv_expiry);
-														route_hop.cltv_expiry_delta as u16 >= chan_update.contents.cltv_expiry_delta
-													},
-													c if c == UPDATE|20 => { // channel_disabled
-														let reported_flags = byte_utils::slice_to_be16(&err_packet.failuremsg[2..2+2]);
-														onion_failure_log!("channel_disabled", UPDATE|20, "flags", reported_flags);
-														chan_update.contents.flags & 0x01 == 0x01
-													},
-													c if c == UPDATE|21 => true, // expiry_too_far
-													_ => { unreachable!(); },
-												};
-
-												let msg = if is_chan_update_invalid { None } else {
-													Some(msgs::HTLCFailChannelUpdate::ChannelUpdateMessage {
-														msg: chan_update,
-													})
-												};
-												res = Some((msg, true));
-												return;
-											}
+							let mut fail_channel_update = None;
+
+							if error_code & NODE == NODE {
+								fail_channel_update = Some(msgs::HTLCFailChannelUpdate::NodeFailure { node_id: route_hop.pubkey, is_permanent: error_code & PERM == PERM });
+							}
+							else if error_code & PERM == PERM {
```

Comment:
What we should do for the BADONIONs isn't entirely clear here (we should amend the spec). I *think* the most clear solution is:
a) change decode_update_add_htlc_onion to always return update_fail_malformed_htlc instead of update_fail_htlc in case of BADONION cases, as even if we get a valid public key, if the HMAC fails, we aren't sure that we got the *right* public key, and should probably just pass it back for the previous node to fail,
b) blame the next-hop channel here (as we do for 8|9|10) instead of the incoming channel.

Note that if we do this we actually end up always blaming the next-hop channel.